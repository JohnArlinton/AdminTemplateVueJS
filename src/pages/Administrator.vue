<template>
  <v-card>
    <v-card-title>
      <v-col cols="12" sm="5" class="order-search">
        <v-text-field
          v-model="search"
          append-icon="mdi-magnify"
          label="Buscar"
          single-line
          hide-details
        ></v-text-field>
      </v-col>
      <v-spacer></v-spacer>
      <v-col cols="12" sm="2" class="text-right order-icon">
        <v-btn
          color="primary"
          ref="addUser"
          dark
          class="mr-5"
          @click="dialog =  true"
          icon
          large
        ><v-icon>person_add</v-icon></v-btn>
        <v-btn icon small @click="getUsers()"><v-icon color="primary">refresh</v-icon></v-btn>
      </v-col>
    </v-card-title>

    <v-data-table
      :headers="headers"
      ref="dataTableUsers"
      :items="users"
      :search="search"
      sort-by="calories"
      class="elevation-1"
    >
      <template v-slot:top>
        <v-toolbar flat color="white">
          <v-spacer></v-spacer>
          <v-dialog v-model="dialog" max-width="700px">
            <v-card style="padding: 20px 0px">
              <v-card-title class="justify-center">
                <span class="headline">{{ formTitle }}</span>
              </v-card-title>

              <v-card-text>
                <v-container>
                  <v-form ref="form" v-model="valid" class="row">
                    <v-col cols="12" md="6">
                      <v-text-field ref="firstName" @keyup.enter="save" v-model="editedItem.firstName" :rules="firstNameRules" label="Name" required></v-text-field>
                    </v-col>
                    <v-col cols="12" md="6">
                      <v-text-field ref="lastName" @keyup.enter="save" v-model="editedItem.lastName" :rules="lastNameRules" label="LastName" required></v-text-field>
                    </v-col>
                    <v-col cols="12" md="6">
                      <v-text-field ref="email" @keyup.enter="save" v-model="editedItem.email" :rules="emailRules" label="Email" required></v-text-field>
                    </v-col>
                    <v-col cols="12" md="6">
                      <v-select ref="role" @keyup.enter="save" v-model="editedItem.roleId" label="Role" :items="roles" item-text="name" item-value="id" required></v-select>
                    </v-col>
                    <v-col cols="12" md="6" class="text-right">
                      <v-checkbox ref="active" color="green" v-model="editedItem.active" label="Activate" hide-details required></v-checkbox>
                    </v-col>
                  </v-form>
                </v-container>
              </v-card-text>

              <v-card-actions class="justify-center">
                <v-btn large dark color="red darken-1 mr-15" @click="close">Cancel</v-btn>
                <v-btn ref="submit" large dark color="green darken-1" :disabled="!valid" @click="save">sAVE</v-btn>
              </v-card-actions>
            </v-card>
          </v-dialog>
        </v-toolbar>
      </template>

      <template v-slot:item.fullName="{ item }">
        {{item.firstName + ' ' +item.lastName}}
      </template>

      <template v-slot:[`item.active`]="{ item }">
        <v-icon :color="getColor(item.active)" dark>{{ getActive(item.active) }}</v-icon>
      </template>

      <template v-slot:item.actions="{ item }">
        <v-btn icon @click="editItem(item)" id="editUser" ref="editUser">
          <v-icon
            small
            class="mr-2"
          >
            mdi-pencil
          </v-icon>
        </v-btn>
        <v-icon
          small
          class="mr-2"
          @click="deleteItem(item)"
        >
          delete
        </v-icon>
      </template>
    </v-data-table> 
  </v-card>
  
</template>

<script>
  import { mapState } from 'vuex'

  export default {
    created() {
      this.getUsers()
      this.$store.dispatch('GET_ROLES')
    },
    data: () => ({
      dialog: false,
      valid: true,
      items:[
        {
          fullName: 'John Doe',
          email: 'johndoe@gmail.com',
          roleName: 'Administrator',
          active: false,
        },
        {
          fullName: 'Alejandro Test',
          email: 'alejandrotest@gmail.com',
          roleName: 'Administrator',
          active: true,
        },
      ],
      firstNameRules: [
        v => !!v || 'El nombre es requerido'
      ],
      lastNameRules: [
        v => !!v || 'El apellido es requerido'
      ],
      emailRules: [
        v => !!v || 'El correo es requerido',
        v => /.+@.+\..+/.test(v) || 'El correo debe ser valido',
      ],
      search: '',
      headers: [
        {
          text: 'Full Name',
          align: 'start',
          value: 'fullName',
        },
        { text: 'Email', value: 'email' },
        { text: 'Rol', value: 'roleName' },
        { text: 'Active', value: 'active' },
        { text: 'Actions', value: 'actions', sortable: false, align: 'center' },
      ],
      editedIndex: -1,
      editedItem: {},
      defaultItem: {
        id: 0,
        firstName: '',
        lastName: '',
        user: '',
        email: '',
        roleName: '',
        // active: false,
      },
    }),

    computed: {
      ...mapState(['users', 'roles',]),
      formTitle () {
        return this.editedIndex === -1 ? 'Create Use' : 'Edit User'
      },
    },

    watch: {
      dialog (val) {
        val || this.close()
        if(this.valid==false){
          this.$refs.form.resetValidation()
          //this.editedItem.roleName = 'Administrator'
        }
      },
    },

    methods: {
      getUsers (){
        this.$store.dispatch('GET_USERS')
      },
      getActive (active) {
        if (active == true) return 'check_circle'
        else return 'remove_circle'
      },
      getColor (active) {
        if (active == true) return 'green'
        else return 'red'
      },

      deleteItem(item) {
        let user = Object.assign({}, item)
        this.$store.dispatch('DELETE_USER', user)
      },

      editItem (item) {
        this.editedIndex = this.users.indexOf(item)
        this.editedItem = Object.assign({}, item)
        this.dialog = true
      },

      close () {
        this.dialog = false
        this.$nextTick(() => {
          this.editedItem = {}
          this.editedIndex = -1
        })
      },

      save () {
        if(this.valid){
          let user = Object.assign({}, this.editedItem)
          user.id = this.editedIndex > -1 ? user.id : 0
          if(this.editedIndex > -1){
            this.$store.dispatch('PUT_USER', user)
          }else{
            this.$store.dispatch('POST_USER', user)
          }
          this.close()
        }
      },
    },
  }
</script>

<style scoped>
  .darken-1{
    margin: 0px 10px;
  }
  @media (max-width: 600px){
  .order-search{
   order: 2; 
  }
  .order-icon{
    order: 1;
  }
}
</style>