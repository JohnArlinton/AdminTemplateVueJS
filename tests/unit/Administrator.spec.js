import Administrator from '../../src/pages/Administrator.vue';
import { shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import Vuetify from 'vuetify';
import Vue from 'vue';

Vue.use(Vuetify, {})
Vue.use(Vuex)

// localVue.config.productionTip = false;

describe('Adminitrator.vue', () => {
  let store, actions

  beforeEach(() =>{
    actions = {
      POST_USER: jest.fn(),
      DELETE_USER: jest.fn(),
      CLICK_LINK: jest.fn(),
      CLICK_IMG: jest.fn(),
    }
  
    const mutations = {
      testMutation: jest.fn()
    }
    
    store = new Vuex.Store({ mutations, actions })
  })
  
  const build = () => {
    const wrapper = shallowMount(Administrator, {
      store,
      Vue
    })
    
    return {
      wrapper,
      dataTableUsers: () => wrapper.findComponent({ref: 'dataTableUsers'}),
      inputFirstName: () => wrapper.findComponent({ref: 'firstName'}),
      inputLastName: () => wrapper.findComponent({ref: 'lastName'}),
      inputEmail: () => wrapper.findComponent({ref: 'email'}),
      selectRole: () => wrapper.findComponent({ref: 'role'}),
      checkActive: () => wrapper.findComponent({ref: 'active'}),
      buttonAddUser: () => wrapper.findComponent({ref: 'addUser'}),
      buttonEditUser: () => wrapper.findComponent({ref: 'editUser'}),
      buttonSubmit: () => wrapper.findComponent({ref: 'submit'}),
    }
  }

  it('it renders the component', async () => {
    const { wrapper, dataTableUsers, inputFirstName, inputLastName, inputEmail, selectRole, checkActive, buttonAddUser, buttonEditUser, buttonSubmit } = build()

    
    // wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()

    console.log(wrapper.vm.$data)
    expect(buttonAddUser().exists()).toBe(true)

    await buttonAddUser().trigger("click")

    console.log(wrapper.vm.$data.dialog)

    expect(wrapper.vm.$data.dialog).toBe(true)

    expect(inputFirstName().exists()).toBe(true)
  })
})