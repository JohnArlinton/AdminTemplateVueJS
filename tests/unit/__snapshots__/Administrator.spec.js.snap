// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adminitrator.vue ir renders the component 1`] = `
<v-card>
  <v-card-title>
    <v-col cols="12" sm="5" class="order-search">
      <v-text-field append-icon="mdi-magnify" label="Buscar" single-line="" hide-details=""></v-text-field>
    </v-col>
    <v-spacer></v-spacer>
    <v-col cols="12" sm="2" class="text-right order-icon">
      <v-btn color="primary" dark="" icon="" large="" class="mr-5">
        <v-icon>person_add</v-icon>
      </v-btn>
      <v-btn icon="" small="">
        <v-icon color="primary">refresh</v-icon>
      </v-btn>
    </v-col>
  </v-card-title>
  <v-data-table headers="[object Object],[object Object],[object Object],[object Object],[object Object]" search="" sort-by="calories" class="elevation-1"></v-data-table>
</v-card>
`;

exports[`Adminitrator.vue it renders the component 1`] = `
<v-card-stub loaderheight="4" tag="div">
  <v-card-title-stub>
    <v-col-stub cols="12" sm="5" tag="div" class="order-search">
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" appendicon="mdi-magnify" backgroundcolor="" hidedetails="true" label="Buscar" loaderheight="2" clearicon="$clear" singleline="true" type="text"></v-text-field-stub>
    </v-col-stub>
    <v-spacer-stub></v-spacer-stub>
    <v-col-stub cols="12" sm="2" tag="div" class="text-right order-icon">
      <v-btn-stub color="primary" dark="true" tag="button" activeclass="" large="true" icon="true" type="button" class="mr-5">
        <v-icon-stub>person_add</v-icon-stub>
      </v-btn-stub>
      <v-btn-stub tag="button" activeclass="" small="true" icon="true" type="button">
        <v-icon-stub color="primary">refresh</v-icon-stub>
      </v-btn-stub>
    </v-col-stub>
  </v-card-title-stub>
  <v-data-table-stub mobilebreakpoint="600" items="" options="[object Object]" sortby="calories" sortdesc="" customsort="[Function]" page="1" itemsperpage="10" groupby="" groupdesc="" customgroup="[Function]" locale="en-US" search="" customfilter="[Function]" serveritemslength="-1" itemkey="id" value="" expanded="" noresultstext="$vuetify.dataIterator.noResultsText" nodatatext="$vuetify.noDataText" loadingtext="$vuetify.dataIterator.loadingText" selectablekey="isSelectable" loaderheight="4" headers="[object Object],[object Object],[object Object],[object Object],[object Object]" expandicon="$expand" itemclass="" class="elevation-1"></v-data-table-stub>
</v-card-stub>
`;
